File: ruby/lib/runapi/kling/resources/text_to_video.rb
# frozen_string_literal: true

module RunApi
  module Kling
    module Resources
      # Kling text-to-video resource.
      # Generate videos from text prompts.
      class TextToVideo
        include RunApi::Core::ResourceHelpers
        include O1ReferenceValidation

        ENDPOINT = "/api/v1/kling/text_to_video"

        RESPONSE_CLASS = Types::TextToVideoResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedTextToVideoResponse
        V26_MODEL = "kling-v2.6"
        V3_TURBO_MODEL = "kling-v3-turbo-text-to-video"
        V3_TURBO_UNSUPPORTED_FIELDS = %i[
          enable_sound
          negative_prompt
          cfg_scale
          multi_shots
          multi_prompt
          first_frame_image_url
          last_frame_image_url
          kling_elements
        ].freeze

        def initialize(http)
          @http = http
        end

        # Generate a text-to-video task and wait until complete.
        #
        # @param params [Hash] text-to-video parameters
        # @return [RunApi::Kling::Types::CompletedTextToVideoResponse] completed task with videos
        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        # Create a text-to-video task.
        #
        # @param params [Hash] text-to-video parameters
        # @return [RunApi::Kling::Types::TextToVideoResponse] task creation result with id
        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        # Get text-to-video task status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::Kling::Types::TextToVideoResponse] current task status
        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          reject_unsupported_v3_turbo_fields!(params)
          validate_contract!(CONTRACT["text-to-video"], params)
          validate_v26_params!(params)
          validate_kling_o1_references!(params)

          # Bespoke cross-field rules the contract cannot express.
          multi_shots = param(params, :multi_shots) == true

          if multi_shots
            raise Core::ValidationError, "enable_sound must be true when multi_shots is true" unless param(params, :enable_sound) == true
            raise Core::ValidationError, "last_frame_image_url is not supported when multi_shots is true" if param(params, :last_frame_image_url)

            multi_prompt = param(params, :multi_prompt)
            validate_multi_prompt!(multi_prompt)
          else
            raise Core::ValidationError, "prompt is required" unless param(params, :prompt)
          end
        end

        def validate_v26_params!(params)
          return unless param(params, :model) == V26_MODEL
          return unless param(params, :enable_sound) == true && param(params, :mode) != "pro"

          raise Core::ValidationError, "enable_sound requires mode pro for #{V26_MODEL}"
        end

        def reject_unsupported_v3_turbo_fields!(params)
          return unless param(params, :model) == V3_TURBO_MODEL

          field = V3_TURBO_UNSUPPORTED_FIELDS.find { |candidate| field_present?(params, candidate) }
          raise Core::ValidationError, "#{field} is not supported by #{V3_TURBO_MODEL}" if field
        end

        def validate_multi_prompt!(multi_prompt)
          unless multi_prompt.is_a?(Array) && multi_prompt.any?
            raise Core::ValidationError, "multi_prompt must be a non-empty array when multi_shots is true"
          end

          multi_prompt.each_with_index do |shot, index|
            prompt = shot.is_a?(Hash) ? (shot[:prompt] || shot["prompt"]) : nil
            duration_seconds = shot.is_a?(Hash) ? (shot[:duration_seconds] || shot["duration_seconds"]) : nil

            raise Core::ValidationError, "multi_prompt[#{index}].prompt is required" if prompt.nil? || prompt.empty?

            if prompt.length > Types::MULTI_PROMPT_MAX_LENGTH
              raise Core::ValidationError, "multi_prompt[#{index}].prompt exceeds #{Types::MULTI_PROMPT_MAX_LENGTH} characters"
            end

            raise Core::ValidationError, "multi_prompt[#{index}].duration_seconds is required" if duration_seconds.nil?

            dur_int = duration_seconds.to_i
            unless Types::MULTI_PROMPT_DURATION_RANGE.cover?(dur_int)
              raise Core::ValidationError, "multi_prompt[#{index}].duration_seconds must be between #{Types::MULTI_PROMPT_DURATION_RANGE.min} and #{Types::MULTI_PROMPT_DURATION_RANGE.max}"
            end
          end
        end
      end
    end
  end
end

File: python/src/runapi/kling/resources/text_to_video.py
"""Kling text-to-video resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from runapi.core import Resource, ValidationError, RequestOptions

from ..contract_gen import CONTRACT
from ..types import (
    MULTI_PROMPT_DURATION_RANGE,
    MULTI_PROMPT_MAX_LENGTH,
    CompletedTextToVideoResponse,
    TextToVideoResponse,
)
from .o1_reference_validation import validate_kling_o1_references

V26_MODEL = "kling-v2.6"
V3_TURBO_MODEL = "kling-v3-turbo-text-to-video"
V3_TURBO_UNSUPPORTED_FIELDS = (
    "enable_sound",
    "negative_prompt",
    "cfg_scale",
    "multi_shots",
    "multi_prompt",
    "first_frame_image_url",
    "last_frame_image_url",
    "kling_elements",
)


class TextToVideo(Resource):
    """Generate videos from text prompts with Kling models."""

    ENDPOINT = "/api/v1/kling/text_to_video"

    RESPONSE_CLASS = TextToVideoResponse
    COMPLETED_RESPONSE_CLASS = CompletedTextToVideoResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Generate a video from text and poll until it completes.

        Args:
            **params: Text-to-video parameters (model, prompt, ...).

        Returns:
            The completed task with videos.
        """
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a text-to-video task and return immediately with an ``id``.

        Args:
            **params: Text-to-video parameters (model, prompt, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of a text-to-video task.

        Args:
            id: The task id.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)

    def _validate_params(self, params: Dict[str, Any]) -> None:
        self._reject_unsupported_v3_turbo_fields(params)
        self._validate_contract(CONTRACT["text-to-video"], params)
        self._validate_v26_params(params)
        validate_kling_o1_references(params)

        # Bespoke cross-field rules the contract cannot express.
        multi_shots = params.get("multi_shots") is True

        if multi_shots:
            if params.get("enable_sound") is not True:
                raise ValidationError("enable_sound must be true when multi_shots is true")
            if params.get("last_frame_image_url"):
                raise ValidationError("last_frame_image_url is not supported when multi_shots is true")

            self._validate_multi_prompt(params.get("multi_prompt"))
        else:
            if not params.get("prompt"):
                raise ValidationError("prompt is required")

    def _validate_v26_params(self, params: Dict[str, Any]) -> None:
        if (
            params.get("model") == V26_MODEL
            and params.get("enable_sound") is True
            and params.get("mode") != "pro"
        ):
            raise ValidationError(f"enable_sound requires mode pro for {V26_MODEL}")

    def _reject_unsupported_v3_turbo_fields(self, params: Dict[str, Any]) -> None:
        if params.get("model") != V3_TURBO_MODEL:
            return

        for field in V3_TURBO_UNSUPPORTED_FIELDS:
            if self._field_present(params, field):
                raise ValidationError(f"{field} is not supported by {V3_TURBO_MODEL}")

    def _validate_multi_prompt(self, multi_prompt: Any) -> None:
        if not (isinstance(multi_prompt, list) and len(multi_prompt) > 0):
            raise ValidationError("multi_prompt must be a non-empty array when multi_shots is true")

        for index, shot in enumerate(multi_prompt):
            prompt = shot.get("prompt") if isinstance(shot, dict) else None
            duration_seconds = shot.get("duration_seconds") if isinstance(shot, dict) else None

            if prompt is None or len(prompt) == 0:
                raise ValidationError(f"multi_prompt[{index}].prompt is required")

            if len(prompt) > MULTI_PROMPT_MAX_LENGTH:
                raise ValidationError(
                    f"multi_prompt[{index}].prompt exceeds {MULTI_PROMPT_MAX_LENGTH} characters"
                )

            if duration_seconds is None:
                raise ValidationError(f"multi_prompt[{index}].duration_seconds is required")

            try:
                dur_int = int(duration_seconds)
            except (TypeError, ValueError):
                dur_int = None
            if dur_int is None or dur_int not in MULTI_PROMPT_DURATION_RANGE:
                raise ValidationError(
                    f"multi_prompt[{index}].duration_seconds must be between "
                    f"{MULTI_PROMPT_DURATION_RANGE.start} and {MULTI_PROMPT_DURATION_RANGE.stop - 1}"
                )
